### DISCUSSION TITLE
vector search Query with filters

### DISCUSSION THREAD

**Discussion by patelprateek** [qdrant/qdrant#1179] [General]
Thanks for this amazing work.
Can you please provide some details on current design for query planner when we want to do vector search with some filtering constraints, i observed a pr https://github.com/qdrant/qdrant/pull/1003 which might be relevant , but trying to understand a bit on implementation details here
if the cardinality of query was too low , i assume the previous latency were high because scanning hnsw can be expensive ? why not just do brute force ? 
Can you also elaborate a bit on query planner cardinality estimation , how does this really work , how do we estimate cardinality without actually running the filter

**Comment by generall** [comment]
Hi @patelprateek,  thanks for the question. We are not putting those details into the documentation, because details of execution are changing constantly and we wouldn't have a resource to maintain documentation like this. And in my opinion - no documentation is better than outdated documentation. 

To answer your question I would refer you to the code, where this logic is defined:

- Search query planner - https://github.com/qdrant/qdrant/blob/516dcd7020e2f54d91ecdda87e08333b17d85574/lib/segment/src/index/hnsw_index/hnsw.rs#L217
- Carnality estimation - https://github.com/qdrant/qdrant/blob/026bd040b001f1c66e16fc911322f1f182d1cf0f/lib/segment/src/index/query_estimator.rs#L90
